tools/deploy.py

```diff
@@ -0,0 +1,313 @@
+                                print(
+                                    error_color("Hash collision processing:"),
+                                    package_name,
+                                    file=sys.stderr,
+                                )
+                                print(f"  File:        {result.target_path}", file=sys.stderr)
+                                print(f"  Short hash:  {short_mpy_hash}", file=sys.stderr)
+                                print(f"  Full hash:   {mpy_hash}", file=sys.stderr)
+                                with open(output_file_path, "rb") as f:
+                                    print(f"  Target hash: {file_hash(f)}", file=sys.stderr)
+                                print(
+                                    f"Try increasing --hash-prefix (currently {hash_prefix_len})"
+                                )
+                                sys.exit(1)
+                        else:
+                            # Create new file.
+                            ensure_path_exists(output_file_path)
+                            shutil.copyfile(mpy_file.name, output_file_path)
+
+                        # Add the file to the package json.
+                        target_path_mpy = result.target_path[:-2] + "mpy"
+                        package_json["hashes"].append((target_path_mpy, short_mpy_hash))
+
+            # Create/replace {package}/latest.json
+            package_json_path = os.path.join(
+                out_package_dir, str(mpy_version), package_name, "latest.json"
+            )
+            write_json(package_json, package_json_path, minify=True)
+
+            # Write {package}/{version}.json, but only if it doesn't already
+            # exist. A package version is "locked" the first time it's seen
+            # by this script, for the current mpy version.
+            if manifest.metadata().version:
+                versioned_package_json_path = os.path.join(
+                    out_package_dir,
+                    str(mpy_version),
+                    package_name,
+                    f"{manifest.metadata().version}.json",
+                )
+                if not os.path.exists(versioned_package_json_path):
+                    write_json(package_json, versioned_package_json_path, minify=True)
+
+    # Write package index.
+    write_json(index_json, os.path.join(output_path, "index.json"), minify=False)
+
+
+def main():
+    import argparse
+
+    cmd_parser = argparse.ArgumentParser(
+        description="Compile micropython-lib for serving to upip."
+    )
+    cmd_parser.add_argument("--output", required=True, help="output directory")
+    cmd_parser.add_argument("--hash-prefix", default=8, type=int, help="hash prefix length")
+    cmd_parser.add_argument("--mpy-cross", default=None, help="optional path to mpy-cross binary")
+    cmd_parser.add_argument("--micropython", default=None, help="path to micropython repo")
```

Comment:
if you specify the micropython repo, could that be used as the path to mpy-cross?  or will that automatically happen because of line 307 below?

Reply:
Yes. It's a bit complicated because of the two somewhat independent dependencies on mpy_cross and mpy-cross. :scream: 

What I had imagined here is that you'd run this script something like:

- clone micropython repo
- for each bytecode version we want to support:
  - checkout that version of the micropython repo
  - build mpy-cross
  - run deploy.py with --micropython

Reply:
Sounds good.  I tested using just `--micropython` and it all works.